### DISCUSSION TITLE
Help me understand `LinkNeighborLoader(disjoint=True)`

### DISCUSSION THREAD

**Discussion by brovatten** [pyg-team/pytorch_geometric#7900] [Q&A]
Hi, 

I'm using the `LinkNeighborLoader` for link prediction for some edges. To avoid data leakage, I want the two subgraphs generated for a link to be disjoint, meaning there is no path between them. But when I set `disjoint=True`, the returned Data object contains an `edge_label_index` size `(2,2)`. Shouldn't the `edge_label_index` just contain my input link? 

As I use `batch_size=1` and the output of `.batch` is all zeros for every result, I presume the `disjoint` option is for multiple batches, and that `LinkNeighborLoader` guarantees no data leakage, but I wanted to ask here to make sure. 

These are my arguments: 
`LinkNeighborLoader(data=data, edge_label_index=eli, edge_label=el, num_neighbors=[80,40], batch_size=1, is_sorted=True, shuffle=False, disjoint=True)`

Thanks in advance!

**Comment by rusty1s** [comment]
This is currently only possible by ensuring that `data.edge_index` and `data.edge_label_index` do not contain any shared values. You can achieve this via `RandomLinkSplit` for example via the `disjoint_train_ratio` argument.

**Reply by brovatten** [reply]
Hi Matthias. 

I meant the data leakage from the head of a triplet to the tail of a triplet for every sampled subgraph. You are referring to leakage between dataset splits right? Also, what does the additional `edge_label_index` edge in the output mean?

**Reply by brovatten** [reply]
Another thing I'm wondering is how to use mini-batching for non-fixed sizes. Currently I am padding the input to the batch sizes using the nodes with the smallest out degrees as pads. Is there a better way?

**Reply by wgeul** [reply]
I also still struggle a bit with understanding the meaning of `edge_label_index` **that is generated in the output**.

**Reply by wgeul** [reply]
Is this the correct manner to obtain the original `edge_index` that has been sampled in `edge_label_index`?
`torch.take(data.edge_index,edge_label_index)`

**Reply by brovatten** [reply]
Actually no, because the indices you see in `edge_label_index` is are reindexed. The indices in there are mapping to the indices in `n_id` containing the original node ids. You can see here how to retrieve them: 

```
original_nodes = [2, 5, 7, 9] 
reset_edge_index = torch.tensor([[0, 1, 2],[1, 2, 3]])  
mapped_edge_index = torch.tensor(original_nodes)[reset_edge_index]

tensor([[2, 5, 7],
              [5, 7, 9]])
```

**Reply by wgeul** [reply]
@brovatten thanks, that's helpful. I am working with a heterogeneous graph but I think the same principle will apply.

**Reply by brovatten** [reply]
Okay I think I figured it out. Please correct me if I am wrong @rusty1s 

`edge_label_index` will produce double the amount of edges input because it will disjoint both nodes in a link. This means that there now is link in the edge_label_index that connects it with the subgraph of the head, and a second link that connects it with the subgraph of the tail. 

The `disjoint` argument will only guarantee no leakage on a link to link basis. Meaning if we batch several links, there is no guarantee that they information will not leak through subgraphs created by other links in the batch. This needs to be accounted for manually. 

Edit: After examining more I am more confused. How does this input become this output: 

```
tensor([
[ 756111, 1083433, 2155848, 4731623, 5086442],
[1244459,  181071, 2119679,  811872,  323703]])
        
tensor([
[ 756111, 1083433, 2155848, 4731623, 5086442, 1244459,  181071, 2119679, 811872,  323703],
[ 973070, 3498205, 4163093, 3499019, 5348527, 4564240, 3563536, 4720508, 5252297, 4386027]])
```
              
I do notice the first 5 heads in the outputs are heads in the input and the last 5 heads in the outputs are tails in the input. But what does the tails in the output constitute?

**Reply by rusty1s** [reply]
Can you show me an example of how you get this output? You may get an increased size back for `edge_label_index` due to negative sampling, but positive tails shouldn't move to heads instead.

**Reply by rusty1s** [reply]
Please also see https://github.com/pyg-team/pytorch_geometric/pull/7943 where we test that the local and global `edge_label_index` match.
